struct reference mutable function argument builder problem

### Describe the bug

When the struct reference is a mutable function argument it outputs a builder error and a warning.

### Reproduction Steps

```v

@[heap]
struct Client {
mut:
  next &Client
  prev &Client

}

fn init_vm(mut head &Client) {
  	for c := head; c == unsafe{nil}; c = c.next { 

	}
}

fn main() {
	mut head := &Client(unsafe{nil})
   	init_vm(mut head)
}
```

```v

@[heap]
struct Client {
mut:
  next &Client
  prev &Client

}

fn init_vm(mut head &Client) {
  	for c := head; c; c = c.next { 

	}
}

fn main() {
	mut head := &Client(unsafe{nil})
   	init_vm(mut head)
}
```

### Expected Behavior

It not to output a builder error nor a warning.

### Current Behavior

```
issue.v:11:39: warning: cannot assign a reference to a value (this will be an error soon) left=Client false right=Client true ptr=true
    9 | 
   10 | fn init_vm(mut head &Client) {
   11 |       for c := head; c == unsafe{nil}; c = c.next { 
      |                                          ^
   12 | 
   13 |     }
==================
/tmp/v_1000/issue.01HV8TCY7TQM1F0E8ZW6X6ETY7.tmp.c:13024: error: cannot convert 'struct main__Client' to 'unsigned long'
...
==================
(Use `v -cg` to print the entire error message)

builder error
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

 V 0.4.5 423c39c

### Environment details (OS name and version, etc.)

V full version: V 0.4.5 294f7e4.423c39c
OS: linux, Linux version 6.6.13-200.fc39.x86_64 (mockbuild@72f11b2996ed4699b0f705186172808f) (gcc (GCC) 13.2.1 20231205 (Red Hat 13.2.1-6), GNU ld version 2.40-13.fc39) #1 SMP PREEMPT_DYNAMIC Sat Jan 20 18:03:28 UTC 2024
Processor: 4 cpus, 64bit, little endian, Intel(R) Pentium(R) Gold 7505 @ 2.00GHz

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2589,6 +2589,11 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 	mut needs_closing := false
 	if arg.is_mut && !exp_is_ptr {
 		g.write('&/*mut*/')
+	} else if arg.is_mut && arg_typ.is_ptr() && expected_type.is_ptr()
+		&& g.table.sym(arg_typ).kind == .struct_ && expected_type == arg_typ.ref() {
+		g.write('&/*mut*/')
+		g.expr(arg.expr)
+		return
 	} else if exp_is_ptr && !arg_is_ptr && !(arg_sym.kind == .alias
 		&& g.table.unaliased_type(arg_typ).is_pointer() && expected_type.is_pointer()) {
 		if arg.is_mut {

--- a/vlib/v/tests/fn_call_mut_ref_args_test.v
+++ b/vlib/v/tests/fn_call_mut_ref_args_test.v
@@ -1,0 +1,23 @@
+@[heap]
+struct Client {
+mut:
+	next &Client = unsafe { nil }
+	prev &Client = unsafe { nil }
+}
+
+fn init_vm1(mut head &Client) {
+	for c := head; c; c = c.next {
+	}
+}
+
+fn init_vm2(mut head &Client) {
+	for c := head; c == unsafe { nil }; c = c.next {
+	}
+}
+
+fn test_fn_call_mut_ref_args() {
+	mut head := &Client{}
+	init_vm1(mut head)
+	init_vm2(mut head)
+	assert true
+}

--- a/vlib/v/tests/in_expression_test.v
+++ b/vlib/v/tests/in_expression_test.v
@@ -337,6 +337,6 @@ fn in_both_mut_ref(mut arr []&Bar, mut bar &Bar) {
 
 fn test_in_both_mut_and_ref() {
 	mut arr := []&Bar{}
-	mut bar := Bar{}
-	in_both_mut_ref(mut &arr, mut &bar)
+	mut bar := &Bar{}
+	in_both_mut_ref(mut &arr, mut bar)
 }

